python/message.c

```diff
@@ -0,0 +1,1487 @@
+      }
+      PyUpb_WeakMap_Add(self->unset_subobj_map, field, subobj);
+    }
+
+    assert(!PyErr_Occurred());
+    return subobj;
+  }
+
+  if (seq) {
+    assert(!PyUpb_CMessage_IsUnset(self));
+    upb_mutmsgval mutval =
+        upb_msg_mutable(self->msg, field, PyUpb_Arena_Get(self->arena));
+    if (upb_fielddef_ismap(field)) {
+      // TODO(haberman): re-enable when maps are checked in.
+      // return PyUpb_MapContainer_GetOrCreateWrapper(mutval.map, field,
+      //                                              self->arena);
+      (void)mutval;
+      PyErr_SetString(PyExc_NotImplementedError, "access map");
+      return NULL;
+    } else {
+      // TODO(haberman): re-enable when repeated fields are checked in.
+      // return PyUpb_RepeatedContainer_GetOrCreateWrapper(mutval.array, _self,
+      //                                                   field, self->arena);
+      PyErr_SetString(PyExc_NotImplementedError, "access repeated");
+      return NULL;
+    }
+  } else {
+    upb_msgval val;
+    if (PyUpb_CMessage_IsUnset(self)) {
+      // Unset message always returns default values.
+      val = upb_fielddef_default(field);
+    } else {
+      val = upb_msg_get(self->msg, field);
+    }
+    return PyUpb_UpbToPy(val, field, self->arena);
+  }
+}
+
+int PyUpb_CMessage_SetFieldValue(PyObject* _self, const upb_fielddef* field,
+                                 PyObject* value) {
+  PyUpb_CMessage* self = (PyUpb_CMessage*)_self;
+  if (upb_fielddef_issubmsg(field) || upb_fielddef_isseq(field)) {
+    PyErr_Format(PyExc_AttributeError,
+                 "Assignment not allowed to message, map, or repeated "
+                 "field \"%s\" in protocol message object.",
+                 upb_fielddef_name(field));
+    return -1;
+  }
+
+  PyUpb_CMessage_AssureWritable(self);
+
+  upb_msgval val;
+  upb_arena* arena = PyUpb_Arena_Get(self->arena);
+  if (!PyUpb_PyToUpb(value, field, &val, arena)) {
+    return -1;
+  }
+
+  upb_msg_set(self->msg, field, val, arena);
+  return 0;
+}
+
+int PyUpb_CMessage_GetVersion(PyObject* _self) {
+  PyUpb_CMessage* self = (PyUpb_CMessage*)_self;
+  return self->version;
+}
+
+/*
+ * PyUpb_CMessage_GetAttr()
+ *
+ * Implements:
+ *   foo = msg.foo
+ *
+ * Attribute lookup must find both message fields and base class methods like
+ * msg.SerializeToString().
+ */
+__attribute__((flatten)) static PyObject* PyUpb_CMessage_GetAttr(
+    PyObject* _self, PyObject* attr) {
+  PyUpb_CMessage* self = (PyUpb_CMessage*)_self;
+
+  // Lookup field by name.
+  const upb_fielddef* field;
+  if (PyUpb_CMessage_LookupName(self, attr, &field, NULL, NULL)) {
+    return PyUpb_CMessage_GetFieldValue(_self, field);
+  }
+
+  // Check base class attributes.
+  assert(!PyErr_Occurred());
+  PyObject* ret = PyObject_GenericGetAttr(_self, attr);
+
+  // Return value if found, swallow AttributeError if raised.
+  if (ret) {
+    return ret;
+  }
+  if (!PyErr_ExceptionMatches(PyExc_AttributeError)) {
```

Comment:
I think this is a bit more clear as:

```
if (ret) return ret;
if (Matches(AttributeError)) {
  // swallow attribute errors because XXX
  PyErr_Clear();
  return PyUpb...
}

return NULL;
```